src/vanilla/store.ts

```diff
@@ -418,118 +426,156 @@ const buildStore = (
+  const getSortedDependents = (
     pending: Pending,
-    rootAtom: AnyAtom,
-    rootAtomState: AtomState,
-  ): [[AnyAtom, AtomState, number][], Set<AnyAtom>] {
-    const sorted: [atom: AnyAtom, atomState: AtomState, epochNumber: number][] =
-      []
+    rootAtoms: Iterable<AnyAtom>,
+  ) => {
+    const atomMap = getAllDependents(pending, rootAtoms)
+    const sorted: AnyAtom[] = []
     const visiting = new Set<AnyAtom>()
     const visited = new Set<AnyAtom>()
-    // Visit the root atom. This is the only atom in the dependency graph
+    // Visit the root atoms. These are the only atoms in the dependency graph
     // without incoming edges, which is one reason we can simplify the algorithm
-    const stack: [a: AnyAtom, aState: AtomState][] = [[rootAtom, rootAtomState]]
+    const stack: [a: AnyAtom, dependents: Set<AnyAtom>][] = []
+    for (const a of rootAtoms) {
+      if (atomMap.has(a)) {
+        stack.push([a, atomMap.get(a)!])
+      }
+    }
     while (stack.length > 0) {
-      const [a, aState] = stack[stack.length - 1]!
+      const [a, dependents] = stack[stack.length - 1]!
       if (visited.has(a)) {
-        // All dependents have been processed, now process this atom
         stack.pop()
         continue
       }
       if (visiting.has(a)) {
-        // The algorithm calls for pushing onto the front of the list. For
-        // performance, we will simply push onto the end, and then will iterate in
-        // reverse order later.
-        sorted.push([a, aState, aState.n])
+        // The algorithm calls for pushing onto the front of the list.
+        // For performance we push on the end, and will reverse the order later.
+        sorted.push(a)
         // Atom has been visited but not yet processed
         visited.add(a)
         stack.pop()
         continue
       }
       visiting.add(a)
       // Push unvisited dependents onto the stack
-      for (const [d, s] of getMountedOrPendingDependents(pending, a, aState)) {
-        if (a !== d && !visiting.has(d)) {
-          stack.push([d, s])
+      for (const d of dependents) {
+        if (a !== d && !visiting.has(d) && atomMap.has(d)) {
+          stack.push([d, atomMap.get(d)!])
         }
       }
     }
-    return [sorted, visited]
+    return sorted.reverse()
   }
 
-  const recomputeDependents = <Value>(
-    pending: Pending,
-    atom: Atom<Value>,
-    atomState: AtomState<Value>,
-  ) => {
-    // Step 1: traverse the dependency graph to build the topsorted atom list
-    // We don't bother to check for cycles, which simplifies the algorithm.
-    const [topsortedAtoms, markedAtoms] = getSortedDependents(
-      pending,
-      atom,
-      atomState,
-    )
-
-    // Step 2: use the topsorted atom list to recompute all affected atoms
-    // Track what's changed, so that we can short circuit when possible
-    const changedAtoms = new Set<AnyAtom>([atom])
-    for (let i = topsortedAtoms.length - 1; i >= 0; --i) {
```

Comment:
Why not keep using the for-i loop and avoid .reverse()?

Reply:
I didn't like how getSortedDependents returns the opposite of what you would expect, which is a topologically sorted list. It has the same complexity, I think the performance will be comparable.

Reply:
I agree with the readability. I thought the performance matters and we are targeting deep dependencies like the case that recursion doesn't work. Are you sure that the performance is comparable?